File: DFS_BFS/536-minMutations-dfs.py
from math import inf
def minMutations(s,e,bank):
    ans=inf
    def dfs(s,e,d,seen=set()):
        nonlocal ans
        if s==e:
            ans=min(ans,d)
            return True

        if s in seen:
            return False
        
        seen.add(s)
        for mut in "ACGT":
            for i in range(len(s)):
                nei=s[:i]+mut+s[i+1:]
                if nei in bank and nei!=s:
                    dfs(nei,e,d+1,seen)
        seen.remove(s)  #for another path
    
    dfs(s,e,0)
    return ans

s="AACCGGTT"
e="AACCGGTA"
bank=["AACCGGTA"]

print(minMutations(s,e,bank))

File: DFS_BFS/536-minMutations-dfs.cpp
#include<iostream>
#include<string>
#include<set>
#include<vector>
#include<algorithm>  //for std::find

using std::cout;
using STR=std::string;
using SET=std::set<STR>;
using VEC=std::vector<STR>;

bool dfoo(const STR& s,const STR& e, int d,const VEC& bank, SET& seen,int& ans)
{
    if(s==e)
    {
        ans=std::min(ans,d);
        return true;
    }
    if(seen.find(s)!=seen.end())
        return false;
    seen.insert(s);
    STR tmpChars{"ACGT"};
    int n=s.size();
    for(int i=0;i<n;++i)
    {
        for(auto &c:tmpChars)
        {
            STR nei=s.substr(0,i)+c+s.substr(i+1,n);
            if(nei!=s &&(std::find(std::begin(bank),std::end(bank),nei)!=std::end(bank)))
            {
                dfoo(nei,e,d+1,bank,seen,ans);
            }
        }
    }
    seen.erase(s);
}

int minMutations(const STR& s,const STR& e,const VEC& bank)
{
    int ans=INT_MAX;
    SET seen{};
    dfoo(s,e,0,bank,seen,ans);
    cout<<ans;
}

int main()
{
    STR s="AACCGGTT",e="AACCGGTA";
    VEC bank{"AACCGGTA"};
    minMutations(s,e,bank);
}

